Hot folder monitoring

arjansh opened the pull request:
Fixes #1329.

Some things which may need special attention when reviewing this pull request:

- The ScheduleServiceImpl now spawns a `org.apache.commons.io.monitor.FileAlterationMonitor` instance on construction, which spawns a monitoring thread in the initialize method. Please check if the implementation doesn't introduce memory leaks or keeps unwanted references to objects.
- The hot folder monitor is quite trigger happy, when a hot folder trigger is added to as a schedule to a job it instantly starts the job for all initial changes it detects, which may be more than one initially. Please test this and provide some feedback whether or not we want to suppress these first time triggers.
- The Customize schedule dialog in DataCleaner monitor starts with a blue paragraph which describes some of different ways in which you can configure the schedule. It may be desirable to also add the hot folder trigger to this paragraph, which as yet has not been done.

LosD commented on monitor/api/src/main/java/org/datacleaner/monitor/scheduling/model/ScheduleDefinition.java:
```diff
@@ -43,7 +43,8 @@
 	private Map<String,String> _jobMetadataProperties;
-    private boolean _runOnHadoop; 
+    private boolean _runOnHadoop;
+    private String _hotFolder;
```
Not on you, but why in the world does Git/Github think that the _runOnHadoop line has changed?

arjansh replied:
Because it actually has changed (either I or my IDE removed an empty space from the end of the line).

LosD replied:
Oh. The beauty of not printing whitespaces (didn't Github used to highlight the actual change in a darker color?)

LosD commented on monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java:
```diff
@@ -665,4 +703,62 @@ public String getServerDate() {
+        }
+
+        public void onStop(FileAlterationObserver observer) {
+            // Do nothing.
```
Ew. A lot of these. Isn't there an adapter version of FileAlterationListener?

LosD commented:
Two things:
-We'll also need to allow watching for a single file, not only all of the folder. I'd maybe do it by keeping the general structure where the entire folder is watched, and then just filter the filename before doing the invocation.
- Would this also work with a Hadoop execution?

@ClaudiaPHI is probably the one with most insight into running Hadoop from the monitor, maybe a quick review watching for dangers would be nice.

arjansh commented:
@LosD 
- I switched from completely implementing the FileAlterationListener to extending the FileAlterationListenerAdaptor.
- I changed it so you can watch single files.
- I did some testing with Hadoop and I can successfully tigger a job to run on a Hadoop cluster using the hot folder. This does however not imply that I can use a folder/file on hdfs as hot folder.

LosD commented:
No, I wouldn't expect any kind of remote folder watching. It would be super cool, but probably a bit too poll'y (unless they have support in their end, like AWS's long poll etc.)

LosD commented on monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java:
```diff
@@ -385,6 +419,17 @@ else if (triggerType == TriggerType.ONETIME) {
     }
 
+    private FileAlterationObserver createObserver(String fileName) {
+        File file = new File(fileName);
```
final! :)

LosD commented on monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java:
```diff
@@ -369,6 +389,20 @@ else if (triggerType == TriggerType.ONETIME) {
                         _scheduler.scheduleJob(jobDetail, trigger);
                     }
+                } else if (triggerType == TriggerType.HOTFOLDER) {
+                    String hotFolder = schedule.getHotFolder();
```
final

LosD commented on monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java:
```diff
@@ -665,4 +714,40 @@ public String getServerDate() {
     }
+
+    void removeHotFolder(String jobIdentifier) {
+        FileAlterationObserver observer = _registeredHotFolders.get(jobIdentifier);
```
final

LosD commented:
Code looks good, except a few finals.

I know that @ClaudiaPHI is doing a lot of testing now, so I'll be lazy and totally skip that :)

ClaudiaPHI commented:
I have tested and it works fine, but I have some concerns: 

- The job seems to be triggered multiple times. For example: it's triggered once when opening the file, and then after I am saving it. Could we do sometimes to prevent starting jobs every second ? 
- is it ok to assume that if a job is already running do not start again another one ?

arjansh commented:
@ClaudiaPHI 
- Regarding your first comment that the job is triggered when opening a file, did you monitor a folder or a specific file? Because this doesn't happen for me when I monitor a file. This may be caused by shadow files being created on disk when opening a file for editing.
- Regarding your second comment, I don't know if we can assume that. In most cases I would say yes, but maybe some users want to run a job after each individual change to a file, so also is a file is changed again before the current job triggered by a previous change has finished. Also this could mean the final change to a file didn't trigger a job, because an earlier trigger job was already running when the final change was done.

ClaudiaPHI commented:
Wait until release is completed. @LosD will decide when to merge it
